arroyo-connectors/src/websocket.rs

```diff
@@ -0,0 +1,203 @@
+use std::time::Duration;
+
+use anyhow::anyhow;
+use arroyo_rpc::grpc::{
+    self,
+    api::{ConnectionSchema, TestSourceMessage},
+};
+use futures::{SinkExt, StreamExt};
+use tokio::sync::mpsc::Sender;
+use tokio_tungstenite::{connect_async, tungstenite};
+use tonic::Status;
+use typify::import_types;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    pull_opt, serialization_mode, Connection, ConnectionType, EmptyConfig, OperatorConfig,
+};
+
+use super::Connector;
+
+const TABLE_SCHEMA: &str = include_str!("../../connector-schemas/websocket/table.json");
+
+import_types!(schema = "../connector-schemas/websocket/table.json");
+const ICON: &str = include_str!("../resources/sse.svg");
+
+pub struct WebsocketConnector {}
+
+impl Connector for WebsocketConnector {
+    type ConfigT = EmptyConfig;
+
+    type TableT = WebsocketTable;
+
+    fn name(&self) -> &'static str {
+        "websocket"
+    }
+
+    fn metadata(&self) -> grpc::api::Connector {
+        grpc::api::Connector {
+            id: "websocket".to_string(),
+            name: "Websocket".to_string(),
+            icon: ICON.to_string(),
+            description: "Connect to a Websocket server".to_string(),
+            enabled: true,
+            source: true,
+            sink: false,
+            testing: true,
+            custom_schemas: true,
+            connection_config: None,
+            table_config: TABLE_SCHEMA.to_owned(),
+        }
+    }
+
+    fn test(
+        &self,
+        _: &str,
+        _: Self::ConfigT,
+        table: Self::TableT,
+        _: Option<&ConnectionSchema>,
+        tx: Sender<Result<TestSourceMessage, Status>>,
+    ) {
+        tokio::task::spawn(async move {
+            let send = |error: bool, done: bool, message: String| {
```

Comment:
These send calls were a bit hard to figure out until I checked it out. Not a big deal.